File: mvp-java/ui/src/main/java/com/wasisto/githubuserfinder/android/ui/userdetails/UserDetailsActivity.java
package com.wasisto.githubuserfinder.android.ui.userdetails;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.wasisto.githubuserfinder.android.DefaultServiceLocator;
import com.wasisto.githubuserfinder.android.R;
import com.wasisto.githubuserfinder.android.ServiceLocator;
import com.wasisto.githubuserfinder.domain.models.User;
import com.wasisto.githubuserfinder.domain.usecases.GetUserUseCase;

public class UserDetailsActivity extends AppCompatActivity implements UserDetailsView {

    public static final String EXTRA_USERNAME = "username";

    private UserDetailsPresenter presenter;

    private ProgressBar loadingIndicator;

    private ImageView avatarImageView;

    private TextView nameTextView;

    private TextView usernameTextView;

    private TextView companyTextView;

    private TextView locationTextView;

    private TextView blogTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_user_details);

        String username = getIntent().getStringExtra(EXTRA_USERNAME);

        if (username == null) {
            finish();
            return;
        }

        ServiceLocator serviceLocator = DefaultServiceLocator.getInstance(getApplication());

        presenter = new UserDetailsPresenter(
                this,
                new GetUserUseCase(
                        serviceLocator.getGithubDataSource()
                ),
                serviceLocator.getLogger(UserDetailsPresenter.class),
                serviceLocator.getIoExecutorService(),
                serviceLocator.getMainExecutorService()
        );

        loadingIndicator = findViewById(R.id.loadingIndicator);
        avatarImageView = findViewById(R.id.avatarImageView);
        nameTextView = findViewById(R.id.nameTextView);
        usernameTextView = findViewById(R.id.usernameTextView);
        companyTextView = findViewById(R.id.companyTextView);
        locationTextView = findViewById(R.id.locationTextView);
        blogTextView = findViewById(R.id.blogTextView);

        blogTextView.setOnClickListener(v -> presenter.onBlogClick());

        presenter.load(username);
    }

    @Override
    public void showLoadingIndicator() {
        loadingIndicator.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoadingIndicator() {
        loadingIndicator.setVisibility(View.GONE);
    }

    @Override
    public void showUser(User user) {
        Glide.with(this)
                .load(user.getAvatarUrl())
                .placeholder(R.color.teal_200)
                .into(avatarImageView);

        nameTextView.setText(user.getName());
        usernameTextView.setText(user.getLogin());
        companyTextView.setText(user.getCompany());
        locationTextView.setText(user.getLocation());
        blogTextView.setText(user.getBlog());
    }

    @Override
    public void openBrowser(String url) {
        startActivity(
                new Intent(Intent.ACTION_VIEW) {{
                    setData(url.matches("(http|https)://") ? Uri.parse(url) : Uri.parse("http://" + url));
                }}
        );
    }

    @Override
    public void showErrorToast() {
        Toast.makeText(this, R.string.an_error_occurred, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void closeActivity() {
        finish();
    }
}

File: mvp-kotlin/ui/src/main/java/com/wasisto/githubuserfinder/android/ui/userdetails/UserDetailsActivity.kt
package com.wasisto.githubuserfinder.android.ui.userdetails

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.lifecycleScope
import com.bumptech.glide.Glide
import com.wasisto.githubuserfinder.android.DefaultServiceLocator
import com.wasisto.githubuserfinder.android.R
import com.wasisto.githubuserfinder.domain.models.User
import com.wasisto.githubuserfinder.domain.usecases.GetUserUseCase
import kotlinx.android.synthetic.main.activity_user_details.*
import kotlinx.coroutines.Dispatchers

class UserDetailsActivity : AppCompatActivity(), UserDetailsView {

    companion object {
        const val EXTRA_USERNAME = "username"
    }

    private lateinit var presenter: UserDetailsPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_user_details)

        val username = intent.getStringExtra(EXTRA_USERNAME)

        if (username == null) {
            finish()
            return
        }

        val serviceLocator = DefaultServiceLocator.getInstance(application)

        presenter = UserDetailsPresenter(
            this,
            GetUserUseCase(
                serviceLocator.githubDataSource
            ),
            serviceLocator.getLogger(UserDetailsPresenter::class),
            serviceLocator.ioDispatcher,
            serviceLocator.mainDispatcher,
            lifecycleScope
        )

        blogTextView.setOnClickListener { presenter.onBlogClick() }

        presenter.load(username)
    }

    override fun showLoadingIndicator() {
        loadingIndicator.visibility = View.VISIBLE
    }

    override fun hideLoadingIndicator() {
        loadingIndicator.visibility = View.GONE
    }

    override fun showUser(user: User) {
        Glide.with(this)
            .load(user.avatarUrl)
            .placeholder(R.color.teal_200)
            .into(avatarImageView)

        nameTextView.text = user.name
        usernameTextView.text = user.login
        companyTextView.text = user.company
        locationTextView.text = user.location
        blogTextView.text = user.blog
    }

    override fun openBrowser(url: String) {
        startActivity(
            Intent(Intent.ACTION_VIEW).apply {
                data = if (url.matches("(http|https)://".toRegex())) Uri.parse(url) else Uri.parse("http://$url")
            }
        )
    }

    override fun showErrorToast() {
        Toast.makeText(this, R.string.an_error_occurred, Toast.LENGTH_SHORT).show()
    }

    override fun closeActivity() {
        finish()
    }
}